draft-pda-protocol.md

```diff
@@ -573,116 +612,288 @@ PA protocol. The structure of the leader share is similarly protocol specific:
+HPKE config id. Let `[helper]` denote `PACollectReq.helper_url`. The leader
+sends a POST request to `[helper]/aggregate` with the following message:
 
 ~~~
 struct {
   PATask task;
-  uint8 hpke_config_id;
-  PAVerifyReq seq<1..2^24-1>;
-} PAVerifyReqSeq;
+  uint8 helper_hpke_config_id;
+  opaque helper_state<0..2^16>;
+  PAAggregateSubReq seq<1..2^24-1>;
+} PAAggregateReq;
 ~~~
 
-The structure contains the PA task, the HPKE config id, and a sequence of
-*sub-requests*, each corresponding to a unique client report. Sub-requests are
-structured as follows:
+The structure contains the PA task, the helper's HPKE config id, an opaque
+*helper state* string, and a sequence of *sub-requests*, each corresponding to a
+unique client report.  Sub-requests are structured as follows:
 
 ~~~
 struct {
-  opaque enc<1..2^16-1>;
+  opaque enc<0..2^16-1>;
+  uint64 time; // UNIX timestamp (seconds) of the report.
   PAProto proto;
-  select (PAVerifyReq.proto) {
+  select (PAAggregateSubReq.proto) {
     case prio:
-      PrioHelperShare;
-      PrioVerifyReq;
+      PrioHelperInputShare;
+      PrioAggregateSubReq;
     case hits:
-      HitsHelperShare;
-      HitsVerifyReq;
+      HitsHelperInputShare;
+      HitsAggregateSubReq;
   }
-} PAVerifyReq;
+} PAAggregateSubReq;
 ~~~
 
 The `enc` field is the helper's encapsulated HPKE context sent in the report.
-The remainder of the me structure is dedicated to the protocol-specific helper
-share and request parameters used for the current round.
+(This field is marked optional because it is usually only needed for the first
+aggregate request. [TODO: Decide how to mark the helper input share as optional
+as well. This too is usually only needed for the first request.]) The `time`
+field should match the timestamp of the corresponding report. The remainder of
+the structure is dedicated to the protocol-specific helper share and request
+parameters used for the current round.
 
 The helper handles well-formed requests as follows. (As usual, malformed
 requests are handled as described in {{pa-error-common-aborts}}.) It first looks
-for the PA parameters `PAParam` for which `PAVerifyReq.task.id ==
+for the PA parameters `PAParam` for which `PAAggregateReq.task.id ==
 PAParam.task.id`. Next, it looks up the HPKE config and corresponding secret key
-associated with `PAVerifyReq.key_config_id`. If not found, then it aborts and
-alerts the leader with "unrecognized key config". [NOTE: In this situation, the
-leader has no choice but to abort. This falls into the class of error scenarios
-that are addressable by running with multiple helpers.]
+associated with `PAAggregateReq.helper_hpke_config_id`. If not found, then it
+aborts and alerts the leader with "unrecognized key config". [NOTE: In this
+situation, the leader has no choice but to abort. This falls into the class of
+error scenarios that are addressable by running with multiple helpers.]
```

Comment:
Why are you assuming that each report uses the same key_id. Suppose you had a situation in which you had a batch of a day's reports and the helper changed keys halfway through?

Reply:
In that case, the leader would need to split the reports into two requests.

> Why are you assuming that each report uses the same key_id.

It seemed to me like it would be simpler if there's just one HPKE key used per request. 

> Suppose you had a situation in which you had a batch of a day's reports and the helper changed keys halfway through?

The leader would split the reports into two separate requests, one for the old key and another for the new.

Reply:
But that might end up with a tail that could never be aggregated because it was smaller than the minimal batch size.

Reply:
Not true, since the number of "sub-requests" isn't necessarily the same as the batch size. The leader makes several aggregate requests and not just one.

In any case, I'm good with relaxing this constraint. Added a TODO.